kokoro/scripts/build/build_package.sh

```diff
@@ -42,24 +42,54 @@ export_to_sponge_config "PACKAGE_VERSION" "${PKG_VERSION}"
 
 ARTIFACT_REGISTRY="us-docker.pkg.dev"
 sudo docker-credential-gcr configure-docker --registries="${ARTIFACT_REGISTRY}"
-CACHE_LOCATION="${ARTIFACT_REGISTRY}/stackdriver-test-143416/google-cloud-ops-agent-build-cache/ops-agent-cache:${DISTRO}"
+function get_cache_location() {
+  echo "${ARTIFACT_REGISTRY}/stackdriver-test-143416/google-cloud-ops-agent-build-cache/ops-agent-cache$1:${DISTRO}"
+}
+function get_target() {
+  echo "${DISTRO}-build$1"
+}
 
-DOCKER_BUILDKIT=1 docker build . \
-  --cache-from="${CACHE_LOCATION}" \
-  --build-arg BUILDKIT_INLINE_CACHE=1 \
-  --progress=plain \
-  --target "${DISTRO}-build" \
-  -t build_image
+DOCKER_TARGET_SUFFIXES=("" "-diagnostics" "-fluent-bit" "-otel")
+DOCKER_CACHE_FROM=""
+for suffix in "${DOCKER_TARGET_SUFFIXES[@]}"
+do
+  DOCKER_CACHE_FROM="${DOCKER_CACHE_FROM}$(get_cache_location $suffix), "
+done
 
-docker history --no-trunc build_image
+
+function docker_build(){
+  # ,$get_cache_location(-diagnostics),$get_cache_location(-fluent-bit),$get_cache_location(-otel)
+  DOCKER_BUILDKIT=1 docker build . \
+    --cache-from="$(get_cache_location $1),$(get_cache_location -diagnostics),$(get_cache_location -fluent-bit),$(get_cache_location -otel)" \
+    --build-arg BUILDKIT_INLINE_CACHE=1 \
+    --progress=plain \
+    --target "$(get_target $1)" \
+    -t $2
+}
+
+for suffix in "${DOCKER_TARGET_SUFFIXES[@]}"
+do
+  docker pull "$(get_cache_location $suffix)"
+done
+(docker_build "" "build_image")
+
+
+#docker history --no-trunc build_image
 
 # Tell our continuous build to update the cache. Our other builds do not
 # write to any kind of cache, for example a per-PR cache, because the
 # push takes a few minutes and adds little value over just using the continuous
 # build's cache.
+function update_cache() {
+  docker_build "$1" "$(get_cache_location $1)"
```

Comment:
are we rebuilding here? even if it's fully cached, this seems bad.

Reply:
Yes. We need to build each one of the targets individually to be able to update the cache. We could re-use the `build_image` target and upload that, but the remaining targets (`"-diagnostics" "-fluent-bit" "-otel"`) would still require a docker build step for us to be able to upload them.

Is the concern that this step would take too long? If so I can time it as I believe it wouldn't take that long since we already built the main target

Reply:
Let's time it

Reply:
Here are the times

docker build target | time | OS
---------------------------- | ------ | ----
ops-agent-cache-diagnostics | 3.923s | [Bullseye](https://source.cloud.google.com/results/invocations/a5e36519-2184-4dfb-b859-0ee083967c35/log)
ops-agent-cache-fluent-bit | 3.273s | Bullseye
ops-agent-cache-otel | 2.123s | Bullseye
 | | 
ops-agent-cache-diagnostics | 4.245s | [Buster](https://source.cloud.google.com/results/invocations/2b6016ea-db97-45e5-b413-0d7c130bcba0/log)
ops-agent-cache-fluent-bit | 2.758s | Buster
ops-agent-cache-otel | 1.510s | Buster
 | | 
ops-agent-cache-diagnostics | 2.711s | [centos7](https://source.cloud.google.com/results/invocations/56e2588c-6de0-431a-8dbf-fa4afed8b692/log)
ops-agent-cache-fluent-bit | 2.252s | centos7
ops-agent-cache-otel | 1.634s | centos7
 | | 
ops-agent-cache-diagnostics | 4.615s |[rockylinux9](https://source.cloud.google.com/results/invocations/45ec6b6c-3fee-4959-b190-0b10bb1c5afd/targets/stackdriver_agents%2Ftesting%2Fconsumer%2Fops_agent%2Fbuild%2Frockylinux9)
ops-agent-cache-fluent-bit | 2.813s |rockylinux9
ops-agent-cache-otel | 1.678s |rockylinux9
 | | 
ops-agent-cache-diagnostics | 2.026s | [sles15](https://source.cloud.google.com/results/invocations/07d87882-a76f-4bdf-ac35-2e380292ece7/targets/stackdriver_agents%2Ftesting%2Fconsumer%2Fops_agent%2Fbuild%2Fsles15)
ops-agent-cache-fluent-bit | 2.181s | sles15
ops-agent-cache-otel | 1.447s | sles15